src/sentry/tasks/weekly_reports.py

```diff
@@ -0,0 +1,803 @@
+
+
+# The entry point. This task is scheduled to run every week.
+@instrumented_task(
+    name="sentry.tasks.weekly_reports.schedule_organizations",
+    queue="reports.prepare",
+    max_retries=5,
+    acks_late=True,
+)
+def schedule_organizations(dry_run=False, timestamp=None, duration=None):
+    if timestamp is None:
+        # The time that the report was generated
+        timestamp = to_timestamp(floor_to_utc_day(timezone.now()))
+
+    if duration is None:
+        # The total timespan that the task covers
+        duration = ONE_DAY * 7
+
+    organizations = Organization.objects.filter(status=OrganizationStatus.ACTIVE)
+    for i, organization in enumerate(
+        RangeQuerySetWrapper(organizations, step=10000, result_value_getter=lambda item: item.id)
+    ):
+        if features.has("organizations:weekly-email-refresh", organization):
+            # Create a celery task per organization
+            prepare_organization_report.delay(timestamp, duration, organization.id, dry_run=dry_run)
+
+
+# This task is launched per-organization.
+@instrumented_task(
+    name="sentry.tasks.weekly_reports.prepare_organization_report",
+    queue="reports.prepare",
+    max_retries=5,
+    acks_late=True,
+)
+def prepare_organization_report(
+    timestamp, duration, organization_id, dry_run=False, target_user=None
+):
+    organization = Organization.objects.get(id=organization_id)
+    ctx = OrganizationReportContext(timestamp, duration, organization)
+
+    # Run organization passes
+    with sentry_sdk.start_span(op="weekly_reports.user_project_ownership"):
+        user_project_ownership(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.project_event_counts_for_organization"):
+        project_event_counts_for_organization(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.organization_project_issue_summaries"):
+        organization_project_issue_summaries(ctx)
+
+    with sentry_sdk.start_span(op="weekly_reports.project_passes"):
+        # Run project passes
+        for project in organization.project_set.all():
+            project_key_errors(ctx, project)
+            project_key_transactions(ctx, project)
+            project_key_performance_issues(ctx, project)
+
+    with sentry_sdk.start_span(op="weekly_reports.fetch_key_error_groups"):
+        fetch_key_error_groups(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.fetch_key_performance_issue_groups"):
+        fetch_key_performance_issue_groups(ctx)
+
+    # Finally, deliver the reports
+    with sentry_sdk.start_span(op="weekly_reports.deliver_reports"):
+        deliver_reports(ctx, dry_run=dry_run, target_user=target_user)
+
+
+# Organization Passes
+def user_project_ownership(ctx):
+    for (project_id, user_id) in OrganizationMember.objects.filter(
+        organization_id=ctx.organization.id, teams__projectteam__project__isnull=False
```

Comment:
We are only looking at members that have a team associated to it since the member only belongs to a project via their membership in a team; correct?

Reply:
That's right. I'll add some comments for this.